File: examples/sidecar-js/decision-analytics/test/latency-histogram.test.ts
import { describe, expect, it } from 'vitest';

import { LatencyHistogram } from '../src/latency-histogram.js';

describe('LatencyHistogram', () => {
  it('returns zeros when empty', () => {
    const histogram = new LatencyHistogram();
    expect(histogram.count).toBe(0);
    expect(histogram.average).toBe(0);
    expect(histogram.p95).toBe(0);
    expect(histogram.p99).toBe(0);
    expect(histogram.min).toBe(0);
    expect(histogram.max).toBe(0);
  });

  it('uses the single value for average and percentiles', () => {
    const histogram = new LatencyHistogram();
    histogram.add(250);
    expect(histogram.average).toBe(250);
    expect(histogram.p95).toBe(250);
    expect(histogram.p99).toBe(250);
    expect(histogram.min).toBe(250);
    expect(histogram.max).toBe(250);
  });

  it('computes percentile boundaries for one hundred values', () => {
    const histogram = new LatencyHistogram();
    const values = Array.from({ length: 100 }, (_unused, index) => index + 1);
    histogram.addAll(values);

    expect(histogram.count).toBe(100);
    expect(histogram.average).toBe(50.5);
    expect(histogram.p95).toBe(95);
    expect(histogram.p99).toBe(99);
    expect(histogram.min).toBe(1);
    expect(histogram.max).toBe(100);
  });

  it('reports correct min and max', () => {
    const histogram = new LatencyHistogram();
    histogram.addAll([10, 50, 30]);
    expect(histogram.min).toBe(10);
    expect(histogram.max).toBe(50);
  });
});

File: examples/plugins/business_rules/decision_analytics/test/latency_histogram_test.exs
defmodule Examples.BusinessRules.DecisionAnalytics.LatencyHistogramTest do
  use ExUnit.Case, async: true

  alias Examples.BusinessRules.DecisionAnalytics.LatencyHistogram

  test "returns zeros when empty" do
    histogram = LatencyHistogram.new()

    assert LatencyHistogram.count(histogram) == 0
    assert LatencyHistogram.average(histogram) == 0
    assert LatencyHistogram.p95(histogram) == 0
    assert LatencyHistogram.p99(histogram) == 0
    assert LatencyHistogram.min(histogram) == 0
    assert LatencyHistogram.max(histogram) == 0
  end

  test "uses the single value for average and percentiles" do
    histogram = LatencyHistogram.add(LatencyHistogram.new(), 250)

    assert LatencyHistogram.average(histogram) == 250
    assert LatencyHistogram.p95(histogram) == 250
    assert LatencyHistogram.p99(histogram) == 250
    assert LatencyHistogram.min(histogram) == 250
    assert LatencyHistogram.max(histogram) == 250
  end

  test "computes percentile boundaries for one hundred values" do
    values = Enum.to_list(1..100)
    histogram = LatencyHistogram.add_all(LatencyHistogram.new(), values)

    assert LatencyHistogram.count(histogram) == 100
    assert LatencyHistogram.average(histogram) == 50.5
    assert LatencyHistogram.p95(histogram) == 95
    assert LatencyHistogram.p99(histogram) == 99
    assert LatencyHistogram.min(histogram) == 1
    assert LatencyHistogram.max(histogram) == 100
  end

  test "reports correct min and max" do
    histogram = LatencyHistogram.add_all(LatencyHistogram.new(), [10, 50, 30])

    assert LatencyHistogram.min(histogram) == 10
    assert LatencyHistogram.max(histogram) == 50
  end
end
